File: ccc2012s3.py
#Solution by Srihari Vishnu
n = input()
readings = []
freq = {}
for x in range(n):
  a = input()
  readings.append(a)
  if a in freq:
    freq[a] += 1
  else:
    freq[a] = 1
readings.sort()
i =0
high = -1
second  =-1
placesh = []
placess = []

while i < n:
  if (freq[readings[i]] > high):
    second = high
    placess = placesh
    high = freq[readings[i]]
    placesh = [i]
  elif (freq[readings[i]] == high):
    placesh.append(i)
  elif (freq[readings[i]]<high and freq[readings[i]]>second):
    second = freq[readings[i]]
    placess = [i]
  elif (freq[readings[i]] == second):
    placess.append(i)
  
  i+=freq[readings[i]]
if len(placesh) ==1 and len(placess) == 1:
  print abs(readings[placesh[0]]-readings[placess[0]])
elif len(placesh)>1:
  print abs(readings[placesh[0]]-readings[placesh[-1]])
else:
  ans = abs(readings[placesh[0]]-readings[placess[0]])
  ans2 = abs(readings[placesh[0]]-readings[placess[-1]])
  print max(ans, ans2)
  

File: ccc2012s3.cpp
//Solution by Srihari Vishnu
//CCC 2012 S3 : Absolutely Acidic
//Sort and then compare using tie breakers

//CASE 1: Two highest frequencies
//CASE 2: Multiple with highest frequency. 
//CASE 3: Multiple second highest frequency


#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Reading {
  int value;
  int freq;

  bool operator > (const Reading& obj) const {
    if (freq == obj.freq) {
      return (value > obj.value);
    }
    return (freq > obj.freq);
  }
};

int main() {
  int n,x;
  int readings[1001];
  vector<Reading> frequencies;

  fill(readings, readings+1001, 0);

  cin >> n;
  for (int i =0; i < n; i++) {
    cin >> x;
    readings[x] += 1;
  }
  for (int t = 0; t < 1001; t++) {
    Reading temp = Reading();
    temp.freq = readings[t];
    temp.value = t;
    frequencies.push_back(temp);
  }
  sort(frequencies.begin(), frequencies.end(), greater<Reading>());

  //CASE 1
  if (frequencies[0].freq != frequencies[1].freq && frequencies[1].freq != frequencies[2].freq) {
    cout << abs(frequencies[0].value-frequencies[1].value);
  }
  //CASE 2
  else if (frequencies[0].freq == frequencies[1].freq) {
    int i = 1;
    while (frequencies[i].freq == frequencies[0].freq) {
      i++;
    }
    i -= 1;
    cout << abs(frequencies[0].value - frequencies[i].value);
  }
  //CASE 3
  else {
    int i = 2;
    while (frequencies[i].freq == frequencies[1].freq) {
      i++;
    }
    i -= 1;
    cout << max(abs(frequencies[0].value - frequencies[i].value), abs(frequencies[0].value - frequencies[1].value));
  }
}
